samples/solution-counters.js

```diff
@@ -0,0 +1,63 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+'use strict';
+const { Firestore, FieldValue } = require('@google-cloud/firestore');
+
+async function main() {
+  // [START increment_counter]
+  function incrementCounter(docRef, numShards) {
+    const shardId = Math.floor(Math.random() * numShards);
+    const shardRef = docRef.collection("shards").doc(shardId.toString());
+    return shardRef.set({ count: FieldValue.increment(1) }, { merge: true });
+  }
+  // [END increment_counter]
+
+  // [START get_count]
+  async function getCount(docRef) {
+    const snapshotList = await docRef.collection("shards").get();
+
+    return snapshotList.docs.reduce((count, documentSnapshot) => {
```

Comment:
This is pretty clever but IMHO not the easiest code to understand in a sample.

Optional.

Reply:
optimize the query for getting count.

Reply:
This makes the code even harder to follow...

What about:
```
const querySnapshot = await docRef.collection("shards").get();
const documents = querySnapshot.docs();

let count = 0;
for (const doc of documents) {
  count += doc.get("count");
}
return count;
```

While "reduce" is nice and works here, it is non-trivial and much harder to understand than a simple loop.